code/modules/vending/_vending.dm

```diff
@@ -1032,30 +1035,47 @@ IF YOU MODIFY THE PRODUCTS LIST OF A MACHINE, MAKE SURE TO UPDATE ITS RESUPPLY C
 						use_power(5)
 						vend_ready = TRUE
 						return TRUE
-					if(account.has_money(S.custom_price))
-						account.adjust_money(-S.custom_price)
-						var/datum/bank_account/owner = private_a
-						if(owner)
-							owner.adjust_money(S.custom_price)
-							SSblackbox.record_feedback("amount", "vending_spent", S.custom_price)
-						vending_machine_input[N] = max(vending_machine_input[N] - 1, 0)
-						S.forceMove(drop_location())
-						if (usr.CanReach(src) && usr.put_in_hands(S))
-							to_chat(usr, "<span class='notice'>You take [S.name] out of the slot.</span>")
-						else
-							to_chat(usr, "<span class='warning'>[capitalize(S.name)] falls onto the floor!</span>")
-						loaded_items--
-						use_power(5)
-						if(last_shopper != REF(usr) || purchase_message_cooldown < world.time)
-							say("Thank you for buying local and purchasing [S]!")
-							purchase_message_cooldown = world.time + 5 SECONDS
-							last_shopper = REF(usr)
-						vend_ready = TRUE
+					//var/datum/bank_account/account = C?.registered_account
+					if(H.spend_cash(S.custom_price))
+						make_purchase(S, H, N)
 						return TRUE
 					else
-						say("You do not possess the funds to purchase this.")
+						var/additional_message = ""
+						if(!C)
+							additional_message += "No ID card found. "
+						if(!C?.registered_account)
+							additional_message += "No account found. "
+						say("[additional_message]Not enough funds to purchase [S.name].")
+						flick(icon_deny,src)
+						vend_ready = TRUE
+						return
 			vend_ready = TRUE
 
+/obj/machinery/vending/custom/proc/make_purchase(var/obj/bought_item, var/mob/living/carbon/human/H, var/N)
+	var/datum/bank_account/owner = private_a
+	if(owner)
+		owner.adjust_money(bought_item.custom_price)
+		var/obj/item/card/id/id_card = H.get_idcard(TRUE)
+		if(id_card)
+			owner.bank_card_talk("[id_card.registered_name] has bought \a [bought_item] for [bought_item.custom_price] credits at your [name]!")
+		else
+			owner.bank_card_talk("[H.name] has bought \a [bought_item] for [bought_item.custom_price] credits at your [name]!")
+	SSblackbox.record_feedback("amount", "vending_spent", bought_item.custom_price)
+	vending_machine_input[N] = max(vending_machine_input[N] - 1, 0)
+	bought_item.forceMove(drop_location())
+	if (usr.CanReach(src) && usr.put_in_hands(bought_item))
+		to_chat(usr, "<span class='notice'>You take [bought_item.name] out of the slot.</span>")
+	else
+		to_chat(usr, "<span class='warning'>[capitalize(bought_item.name)] falls onto the floor!</span>")
+	playsound(src, 'sound/machines/machine_vend.ogg', 50, TRUE, extrarange = -3)
+	loaded_items--
+	use_power(5)
+	if(last_shopper != REF(usr) || purchase_message_cooldown < world.time)
+		say("Thank you for buying local and purchasing [bought_item]!")
+		purchase_message_cooldown = world.time + 5 SECONDS
```

Comment:
this should be changed as COOLDOWN macro. Since you're here, would you change these please?